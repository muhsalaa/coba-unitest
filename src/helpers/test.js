export function hexToRGBString(hex) {
  const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result
    ? `rgb(${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(
        result[3],
        16
      )})`
    : null;
}

export const findByTestAttr = (wrapper, val, elem = '') => {
  return wrapper.find(`${elem}[data-testid='${val}']`);
};
