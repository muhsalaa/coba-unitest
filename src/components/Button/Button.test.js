import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { FaExclamationTriangle } from 'react-icons/fa';

import { Button } from './index';
import { findByTestAttr } from '../../helpers/test';

describe('Button component', () => {
  describe('Component Renders', () => {
    it('Renders correctly and match snapshot', () => {
      const wrapper = mount(<Button />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('Renders text (children) provided', () => {
      const wrapper = shallow(<Button>Hola como estas</Button>);

      expect(wrapper.text()).toBe('Hola como estas');
    });

    it('Renders icon if provided', () => {
      const wrapper = shallow(
        <Button icon={FaExclamationTriangle}>Hola</Button>
      );

      expect(wrapper.text()).toContain('<FaExclamationTriangle />');
    });
  });

  describe('Component Functionality', () => {
    it('should handle click event', () => {
      const spy = jest.fn();
      const wrapper = mount(<Button onClick={() => spy('haha')} />);

      wrapper.simulate('click');
      wrapper.simulate('click');
      wrapper.simulate('click');
      wrapper.simulate('click');

      expect(spy).toHaveBeenCalledTimes(4);
      expect(spy).toHaveBeenCalledWith('haha');
    });

    it('should not clicked when disabled', () => {
      const spy = jest.fn();
      const wrapper = mount(<Button onClick={() => spy('haha')} disabled />);

      wrapper.simulate('click');

      expect(spy).toHaveBeenCalledTimes(0);
    });
  });

  describe('Component style', () => {
    it('should render blue button by default', () => {
      const wrapper = (props) => mount(<Button {...props} />);
      const button = findByTestAttr(wrapper({}), 'qa-button');

      expect(button.props().className).toContain('bg-blue-600');
    });

    it('should render button with color props provided', () => {
      const wrapper = mount(<Button color="red" />);
      const button = findByTestAttr(wrapper, 'qa-button');

      expect(button.props().className).toContain('bg-red-600');
    });

    it('should render button with gray color when disabled', () => {
      const wrapper = mount(<Button disabled />);
      const button = findByTestAttr(wrapper, 'qa-button');

      expect(button.props().className).toContain('bg-gray-500');
    });

    it('should have full width if block props provided', () => {
      const wrapper = mount(<Button block />);
      const button = findByTestAttr(wrapper, 'qa-button');

      expect(button.props().className).toContain('w-full');
    });
  });
});
