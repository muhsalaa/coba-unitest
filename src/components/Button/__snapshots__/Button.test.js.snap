// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Component Renders Renders correctly and match snapshot 1`] = `
<Button
  className=""
  color="blue"
  disabled={false}
  icon={null}
  onClick={[Function]}
>
  <button
    className="text-base sm:text-lg font-fira font-medium py-1 px-2 md:py-2 md:px-4 rounded-md focus:outline-none flex items-center justify-center bg-blue-600 hover:bg-blue-700 text-white"
    data-testid="qa-button"
    disabled={false}
    onClick={[Function]}
  />
</Button>
`;
